### ISSUE TITLE
IGCM compatibility

### ISSUE THREAD

**Issue by CreativeMD** [mezz/JustEnoughItems#1217]
Hey there,
I'm the author of [IGCM](https://minecraft.curseforge.com/projects/ingameconfigmanager), a mod that allows you to remove or add recipes (crafting table, furnace, etc ...) ingame.

In 1.10.2 (if I'm not mistaken) I could simply reload JEI. As this feature was removed I emptied (through a hacky way) the recipe list of certain categories and added the new recipes. Unfortunately that isn't possible either anymore, I can only hide recipes which is not enough and will result in a couple of issues.

I know this is actually not your problem, but I just don't know how to fix it from my end. Maybe you have an idea?

In Regards
CreativeMD

**Comment by mezz** [comment]
For 1.12 you can still use these deprecated methods:
https://github.com/mezz/JustEnoughItems/blob/86f9c00b679a84ee527bf1d95631ba07a3f9c9a1/src/api/java/mezz/jei/api/IRecipeRegistry.java#L171-L187
Currently there are some performance issues with adding and removing recipes from JEI at runtime so it's really discouraged, but it's not game-breaking.
I try not to make breaking changes except on major version updates, so they will work until 1.13.
With 1.13 I am hoping there will be a better way to do it.

**Comment by CreativeMD** [comment]
Ok thanks for your help! 1.12.2 is not designed for ingame modification of the recipe list (due to the forge registry), so I'm not sure if I will continue doing so anyway, for now my hacky ways work.

**Comment by CreativeMD** [comment]
Unfortunately I ran into a couple of issues ...

I managed to remove all recipe in a certain category by getting all wrappers of that category and removing them again:
```java
List<IRecipeWrapper> recipes = ((IRecipeRegistry) JEIHandler.recipeRegistry).getRecipeWrappers((IRecipeCategory) category);

for (IRecipeWrapper recipe : recipes) {
	((IRecipeRegistry) JEIHandler.recipeRegistry).removeRecipe(recipe, getJEICategory());
}
```
As I updated my JEI repository it did not work anymore. Looks like recipes will not be removed anymore ([source](https://github.com/mezz/JustEnoughItems/blob/1.12/src/main/java/mezz/jei/recipes/RecipeRegistry.java#L324)) but only be hidden. This will result in a memory leak and there also seems to be an issue regarding furnace recipe, which will not be hidden at all ([source](https://github.com/mezz/JustEnoughItems/blob/1.12/src/main/java/mezz/jei/recipes/RecipeRegistry.java#L740)).

Furthermore I had a couple of issues adding new recipes. First of all JEI does not recognize the category (therefore `IRecipeRegistry.addRecipe(Object)` does not work), so I have to use private methods via reflection in order to make it working (`RecipeRegistry.addRecipe(Object,String)`) or use the wrapper version instead (so that one is kinda solved). Unfortunately only furnace recipe will show up properly, crafting recipes are not visible, even though the gui will show up. Additionally, some recipes will not be added because there is no wrapper for them (don't know if that's a problem).

**Comment by mezz** [comment]
The old methods have issues, there is a reason they are deprecated and point to later versions. You should only `add` and `hide` (the wrapper versions that take a recipe category), I think they will work properly. They are used  by gamestages and crafttweaker.
Hiding recipes is the performance issue I mentioned, it takes up memory to have hidden recipes around.

There does appear to be a bug with `removeRecipe`, it should call the newer version of `hideRecipe`. I can fix that now.

**Comment by CreativeMD** [comment]
Ok that fixed everything about the furnace issue, but adding crafting recipes still does not work :(

**Comment by mezz** [comment]
Can you show me the code are you using to add crafting recipes? I think other mods are able to add them.

**Comment by CreativeMD** [comment]
The code is spread across different classes, but in essence this is how it looks like:

```java

@Override
public void onRuntimeAvailable(IJeiRuntime jeiRuntime) {
    JEIHandler.recipeRegistry = jeiRuntime.getRecipeRegistry();
}

IRecipeCategory category = ((IRecipeRegistry) JEIHandler.recipeRegistry).getRecipeCategory(VanillaRecipeCategoryUid.CRAFTING);

List<IRecipeWrapper> oldRecipes = ((IRecipeRegistry) JEIHandler.recipeRegistry).getRecipeWrappers(category);
for (IRecipeWrapper recipe : oldRecipes) {
    ((IRecipeRegistry) JEIHandler.recipeRegistry).removeRecipe(recipe, category.getUid());
}

List recipes = new ArrayList<>();
recipes.addAll(CraftingRecipeChecker.getValidRecipes(((IModRegistry) JEIHandler.modRegistry).getJeiHelpers()));
recipes.addAll(TippedArrowRecipeMaker.getTippedArrowRecipes());
for (Object recipe : recipes) {
    IRecipeWrapper wrapper = ((IRecipeRegistry) JEIHandler.recipeRegistry).getRecipeWrapper(recipe, category.getUid());
    if(wrapper != null)
        ((IRecipeRegistry) JEIHandler.recipeRegistry).addRecipe(wrapper, category.getUid());
}
```

![](https://i.imgur.com/NyUSwcu.png)

The recipes will be added, I could not find anything suspicious will debugging it, but they will not be visible ingame, except of the TippedArrowRecipes:
![](https://i.imgur.com/xHi4GQc.png)

I do the same for the furnace and that one works without any problems.

**Comment by mezz** [comment]
I hope you are not recreating the recipes using JEI internals in the real mod like this:
```java
recipes.addAll(CraftingRecipeChecker.getValidRecipes(((IModRegistry) JEIHandler.modRegistry).getJeiHelpers()));
recipes.addAll(TippedArrowRecipeMaker.getTippedArrowRecipes());
```
Those methods are pretty slow.

Aside from that it looks like it should work, I'll try it out and see what's going on.

**Comment by CreativeMD** [comment]
If you want to I can send you a deobfuscated version of IGCM and CreativeCore so you can test it in your dev environment.

**Comment by mezz** [comment]
That would be helpful, thanks.

**Comment by CreativeMD** [comment]
https://www.dropbox.com/s/a9tstgielz3hfe1/CreativeCore%2BIGCM.jar?dl=1

[source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/api/machine/RecipeMachine.java#L189)

You don't need to configure anything. IGCM will remove all crafting recipes and add them back.

**Comment by mezz** [comment]
Ok I figured out the bug, when you add items that are hidden, they are not unhidden. The fix is pretty straightforward and will be in the next release.

I think the tipped arrows still show up because you use JEI internal classes like this:
https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/machines/WorkbenchMachine.java#L370-L371
It creates new instances of those recipes. After I fix this bug, you might see two instances of each tipped arrow recipe. Let me know if it still happens when you are not recreating them, it might be a different bug.

Please stop using JEI internal classes, it causes problems and makes debugging confusing.

**Comment by CreativeMD** [comment]
The problem is that recipes will be added but not removed, meaning there will be multiple copies of a recipes right after joining (the more often the config will change the more duplicates there will be).

**Comment by CreativeMD** [comment]
Is it possible to thread the reloading of JEI?

**Comment by mezz** [comment]
JEI does most of its reloading with a tick handler that spreads the reload over a long time if it takes more than 10 seconds initially.

**Comment by CreativeMD** [comment]
Okay? But why does the client timeout then? Can't you send a `CPacketKeepAlive` packet if it takes too long?

**Comment by mezz** [comment]
In my testing, the client doesn’t time out because of JEI. You are asking me questions backwards without context, it’s getting a bit frustrating. What are you doing? What is happening wrong? Why does it matter? 
Why don’t you reply when I talk about using JEI internals? For all I know you are manually reloading JEI yourself with some internal call, and no I don’t want to troubleshoot that.

**Comment by CreativeMD** [comment]
Sorry for the confusion. Let me explain ...

**JEI interals**
I tried out everything (because I thought I had messed up something), that's why I eventually started to use those two methods (`CraftingRecipeChecker.getValidRecipes()` and `TippedArrowRecipeMaker.getTippedArrowRecipes()`) to make sure I was doing it the same way as JEI does (I didn't clean up the code that's why it was still in there). I removed it as soon as you mentioned it.

**Timeout**
Back in 1.11.2 I could use `reload()`, but it has been "removed" due to that reason:
![](https://i.imgur.com/s0DycJM.png)
It's by far the easiest solution and I was wondering if there is a way to solve it.
I just had the feeling you don't want to mess around with the problem anymore, since you did not respond to my last post. Furthermore the duplication issue appears to require some deeper code rework, 
 so I thought it would be a good idea to find an easier solution.

**About the duplicates**
I guess recipes duplicate because JEI does not remove, but only hides them. Once a recipe will be added, one or multiple old recipes will unhide as well (causing duplicates to appear). You can test it out, if you want to ([download](https://www.dropbox.com/s/a9tstgielz3hfe1/CreativeCore%2BIGCM.jar?dl=1)). As I mentioned before I got rid of the JEI internals ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/machines/WorkbenchMachine.java#L369)).

**Apologize**
I apologize if my posts confused you. That was not my intention. I tried to search for a better solution, since I feel quite uncomfortable with annoying you with my problems. Hope I could explain everything.

**Comment by mezz** [comment]
Thanks for the explanation. I will try to explain what I know as best I can so we are working with the same information.

When I removed `reload()`, it was being called by CraftTweaker and a few other mods during startup. This would make JEI load 4-5 times in a regular pack shortly after connecting to a server and cause timeouts. For that reason, I took it out. I worked with all of those mods to dine workarounds to do what they needed without reloading JEI, although it was hard for CraftTweaker because it could no longer do a refresh with the game open.

Although I took out that call, JEI still reloads when you do a resource reload, and players of extremely large packs (with mods like Gregtech with many items and recipes) would still timeout when changing the language or texture pack. To support that better, I made it so that the reload is spread out over time and no longer causes timeouts.

Even though the timeout issue is resolved, I do not want to add a reload option again. The way it was used before was extremely bad for load times and players. There is no standard time during server connection for mods to do their thing and reload JEI, so it is not really possible to group these reloads together. Even if I could group the reloads, it would still mean loading JEI multiple times every time you play, which is really not ideal.

When you hide a recipe and then add it again, it should be detected by JEI and the original recipes should be un-hidden instead of added a second time. As far as I understand, the issue you ran into with potion recipes was that you were generating a whole new set of recipes that are not `==` to the first set. When you add them again, JEI does not know they are the same as the previous set of potion recipes.

**Comment by CreativeMD** [comment]
Thanks! That explains it to me, but I think there is one thing you misunderstood: All recipes have duplicates (it's not limited to the potion recipes). Every time the config will be saved a new duplicate will appear:

![](https://i.imgur.com/NOG1qWZ.gif)

**How IGCM works**
IGCM follows a simple rule, which should make it easy for other modders (including myself) to add support for it. It only requires three methods to organize the recipe list.
1. Return all available recipes at startup ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/api/machine/RecipeMachine.java#L149))
2. `clear()`, remove all recipes from the list ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/api/machine/RecipeMachine.java#L129))
3. `add()`, add a new recipe to the list ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/api/machine/RecipeMachine.java#L124))

IGCM will organize the recipe list using these methods (nothing else is required).
There is no detection if recipes are equal therefore I tried to add support for JEI the same way. First I try to remove all recipes from the list ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/api/machine/RecipeMachine.java#L191)) and later on add the new ones ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/api/machine/RecipeMachine.java#L196)).

**JEI**
Why can't JEI remove recipes? I don't know the reason why it only hides a recipe (instead of removing it completely). Maybe you can tell me?

Maybe add a clear method which would remove all recipes of a specific category `clear(String categoryId)`. It would also solve the memory issue. I don't know if that is possible, but as I said earlier, I no idea how to fix the duplication issue from my end.

**Comment by mezz** [comment]
I would like to test it, but IGCM on curse doesn't seem to do anything to JEI and there is no way to set up a workspace from the source on Github. Can you add gradle and workspace configuration to github so that I can test it out?

**Comment by mezz** [comment]
It is not very simple to remove recipes from JEI because of the data structures it uses for lookups and searching.

**Comment by CreativeMD** [comment]
I sent you the deofuscated version earlier ([download](https://www.dropbox.com/s/a9tstgielz3hfe1/CreativeCore%2BIGCM.jar?dl=1)). I also uploaded a video some time ago, which explains how I setup my workspace ([view](https://www.youtube.com/watch?v=7Ahshi_QjM4)). If the deofuscated version doesn't work just let me know I can also give you a pre-release.

What about removing an entire category, would that be easier?

**Comment by mezz** [comment]
I'm not sure why you are setting up your workspace manually, it seems like a large barrier to contributions or collaboration.
I assumed the deofuscated version was still using JEI internals. I will try it out.

**Comment by mezz** [comment]
I have pushed a fix that will automatically unhide hidden recipes that are added a second time instead of showing them as duplicated. https://github.com/mezz/JustEnoughItems/commit/a0d9af5b08b530d0c55d8384f5b838e12c3dd2c5

**Comment by CreativeMD** [comment]
Hey there, sorry for not responding earlier, there were quite a few other things to do, but I finally managed to write down everything.

After I stated that there is nothing I can do about it, I looked at the code and tried to implement a clear method (as mentioned above). After some time I managed to do it, here is an overview of the changes:

In Table.java:
```java
public void removeRow(R key)
{
   table.remove(key);
}
```

In MultiMap.java:
```java
public void clearKey(K key)
{
   map.remove(key);
}
```

In RecipeMap.java:
```java
public void clearCategory(IRecipeCategory category)
{
   recipeWrapperTable.removeRow(category);
   for (Iterator<Entry<String, List<String>>> iterator = categoryUidMap.entrySet().iterator(); iterator.hasNext();) {
      Entry<String, List<String>> entry = iterator.next();
      entry.getValue().remove(categoryId);
      if(entry.getValue().isEmpty())
         iterator.remove();
   }
}
```

In RecipeRegistry.java:
```java
public void clearCategory(IRecipeCategory category)
{
   hiddenRecipes.clearKey(category.getUid());
   wrapperMaps.removeRow(category.getUid());
   recipeWrappersForCategories.clearKey(category);
   recipeCategoriesVisibleCache.clear();
}
```

In theory I was going to suggest you to implement it, but as you solved the duplication issue anyway I think this kind of method is not needed anymore (I can still create a pr if you want to).

Thanks for your kind help and support! I'm happy that you do care about cross-mod compatibility!

For now I will do some more testing related to the `clearCategory(IRecipeCategory)` ([source](https://github.com/CreativeMD/IGCM/blob/1.12/src/main/java/com/creativemd/igcm/jei/JEIHandler.java#L104)) method as it solves some other minor issues (such as a small memory issue and empty guis). Also it's faster (performance wise) and easier to use, so maybe I will stick with it (hope you don't mind).

Anyways, thanks again for your kind and fast help! 1.13 will probably change everything anyway. Not sure if IGCM will survive that update (it's quite hacky already).

In kind Regards,
CreativeMD
